Implementing ToolBar

DieterMai opened the pull request:
Implemented the ToolBar and ToolItem widgets.

Most of the features work the same as the native implementation. Some things are different:

### Not Implemented
**Accessibility**: I can not test this so not implemented.
**Touch**: I can not test this so not implemented.
**ToolTips**: This would need a common implementation for all custom widgets. I don't think it is useful to have this implemented separately for every widget.

### Not working
**Not Displayed images**: Some images are not displayed. This seems to be a skija issue. Example is in Snippet373.
**GIFs**: Not supproted. Causes skija to crash
**Not supported Image formats**: Some images have a format that is not supported by skija and cause it to crash. An example for this is the GraficsExample:
```
Exception in thread "main" java.lang.RuntimeException: Failed to makeRaster ImageInfo(_colorInfo=ColorInfo(_colorType=RGBA_8888, _alphaType=UNPREMUL, _colorSpace=null), _width=16, _height=16) [B@617c74e5 16
	at io.github.humbleui.skija.Image.makeRasterFromBytes(Image.java:56)
```

### Different
**Item Resize**: The native ToolItems never change their size. even if the text or the font changes. My implementation does.
**Rescaling of Items**: In the native implementation every ToolItem has the same size as the first one. Images get scaled up/down in this case. I did not do this. Also i was not able to test this because skija refuses to render my custom images :(
**Bugs**: There are some behaviors that seems like bugs to me. I did not try to implement those.
**Animations**: When the background of an item changes, this is actually an animation in the native implementation. I dont know how to animate and it seemed pointless to me.
**Background images**: Background images are tilled in the native implementation. It seems like skija does not provide a method for this. 

### Packages
I Put non-widget classes in a separate package. The reason for this is that i did not want to clutter the widget package with non-widget classes. I hope this is ok.


Also, i did not test this on linux or mac, sorry

DieterMai commented:
Some tests are failing. I will check later.

DieterMai commented:
As far as i see, everything is fixed now.
There are 2 tests that fail which were not failing before, but they fail due to an exception in Skija due to not supported image type:

```
java.lang.UnsupportedOperationException: Unsupported ColorType
	at org.eclipse.swt.graphics.SkijaGC.getSkijaColorType(SkijaGC.java:298)
	at org.eclipse.swt.graphics.SkijaGC.convertSWTImageToSkijaImage(SkijaGC.java:306)
	at org.eclipse.swt.graphics.SkijaGC.drawImage(SkijaGC.java:220)
```

I hope my pull request is still considered.

DieterMai commented:
I just found a bug so i changed this PR to draft.

DieterMai commented:
Finally was able to fix everything.

I had to add a new color type to SkijaGC to fix some tests. This seems to have fixed some of the image issues i saw in manual tests.

Please review.

DenisUngemach commented:
Can you squash all commits into one commit and force push?
This makes the review easier.

DieterMai commented:
Yes, will do

DieterMai commented:
@DenisUngemach I squashed the commits.

DieterMai commented:
@HeikoKlare Thank you for the review. I made the suggested changes. Some notes on that:

> The background color does not reset properly. I just randomly hovered over toolbar items and sometimes it leaves behind a blue (hovering) background and sometimes also an orange one.

This background flickering seems to be an issue with skija. I did not implement any background rendering originally. Somehow if the background is transparent, skija does not provide a consistent result when calling surface.makeImageSnapshot().
Anyway, i fixed it by drawing a background with the background color. As a result, the toolbar is now a bit darker.

> The icons for minimizing and closing part stacks are not drawn correctly. They are different as they are drawn programmatically (in contrast to the others being loaded from files).

The custom drawn images had an unsupported color type. I fixed it by
```
if (redMask == 0xFF && greenMask == 0xFF00 && blueMask == 0xFF0000) {
	return ColorType.UNKNOWN;
}
```
I tried all other ColorTypes. Only UNKNOWN worked.

> Please just remove the old ToolBar and ToolItem classes instead of renaming them.

I did so. As a result, the examples and snippets that originally used the _Old versions are no longer working. This is due to not supported image types (like gif) 

Also: 
When i rebased on master, i realized the custom widgets now extend CustomControl. ToolBar can not do this. ToolBar can be a Parent for other widgets an thus needs to extend Composite for now.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayout.java:
```diff
@@ -0,0 +1,161 @@
+	public static class Row {
+		private static final int NONE = -1;
+
+		final List<ItemRecord> items = new ArrayList<>();
```
Could it be private?

DieterMai replied:
No. Row is used as a return type in ToolBarLayout#rows(). So Row needs to be accessible from outside.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayoutGenerator.java:
```diff
@@ -0,0 +1,165 @@
+import org.eclipse.swt.graphics.*;
+import org.eclipse.swt.widgets.toolbar.ToolBarLayout.*;
+
+public class ToolBarLayoutGenerator {
```
Could it be made package-private?

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayoutGenerator.java:
```diff
@@ -0,0 +1,165 @@
+	private final boolean horizontal;
+	private final boolean wrap;
+	private final boolean shadowOut;
+	private final boolean RTL;
```
I'd stick to the Java naming convention `rtl`.

DieterMai replied:
ok, will do

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayoutGenerator.java:
```diff
@@ -0,0 +1,165 @@
+
+		int maxWidth = offset.x;
+		int maxHeight = offset.y;
+		for(Row row : rows) {
```
A space is missing after `for`.

DieterMai replied:
yikes

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayoutGenerator.java:
```diff
@@ -0,0 +1,165 @@
+			} else {
+				row = new Row(availableSize, offset.x, true);
+				rows.add(row);
+				row.add(itemRecord);
```
Could `row.add(itemRecord)` be moved after the `else {}`?

DieterMai replied:
In general i would like to keep it this way. Each branch does all the work - easy to understand. I understand that some would see it differently.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarLayout.java:
```diff
@@ -0,0 +1,161 @@
+
+import org.eclipse.swt.graphics.*;
+
+public class ToolBarLayout {
```
I wouldn't declare different top-level classes in one file.

DieterMai replied:
??? ToolBarLayout is the only top level class in the file. Am i missing something?

tmssngr replied:
Ah, sorry, my mistake. Looks like I misunderstood GitHub's review UI.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarRenderer.java:
```diff
@@ -0,0 +1,121 @@
+		render(gc, bounds, layout.rows());
+	}
+
+	private void render(GC gc, Rectangle bounds, List<Row> rows) {
```
Why pass `Rectangle bounds` if only its size parts are used? I suggest to pass just `int width, int height`.

DieterMai replied:
Using size instead

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolBarRenderer.java:
```diff
@@ -0,0 +1,121 @@
+
+	private void drawHorizontalSeparator(GC gc, Row row) {
+		int pos = row.position + row.usedSpace.y + 3;
+		gc.setForeground(bar.getDisplay().getSystemColor(SWT.COLOR_WIDGET_NORMAL_SHADOW));
```
Please avoid `display.getSystemColor()`, but rather define a constant. Only then you can be sure that the value is right.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolItemButtonRenderer.java:
```diff
@@ -0,0 +1,357 @@
+
+	// this is package protected so that the ToolItemDropDownRenderer can use the
+	// same implementation
+	void renderHighlight(GC gc, Rectangle bounds) {
```
I'm not sure whether this suites a dark theme, too.

DieterMai replied:
I'm am also not sure, mainly because i did not test it. But for now i would like to focus on one theme and feature parity.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolItemButtonRenderer.java:
```diff
@@ -0,0 +1,357 @@
+		return image;
+	}
+
+	private Image getDisabledImage() {
```
Maybe it makes sense to create an image "container" that covers disabled images, too. This could be used in other controls like `Label` and `Button`, too.

DieterMai replied:
That's a good idea. That would also solve the issue of different implementations. EG Button uses SWT.IMAGE_GRAY as flag for disabled image, but it should be IMAGE_DISABLE.
However, i don't see changes like this in the scope of this PR.

tmssngr replied:
Maybe you can prepare another, smaller pull request before that solves this issue. You then can reuse the "container". Of course, it also can be done later.

DieterMai replied:
I think it would be simpler to do it after

tmssngr commented:
On MacOS also stretching separators were supported, too. They allowed to have toolbar buttons at the left, center and right of the toolbar.

DieterMai commented:
> On MacOS also stretching separators were supported, too. They allowed to have toolbar buttons at the left, center and right of the toolbar.

Good to know. Yet i can not test this and would not know how this works or looks. Maybe someone else could implement it.

DieterMai commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/toolbar/ToolItemButtonRenderer.java:
```diff
@@ -0,0 +1,357 @@
+
+	private Color getTextColor() {
+		if (item.isEnabled()) {
+			return getForegourndColor();
```
fxied

HeikoKlare reviewed (commented):
Thank you for improving the PR and for the in-depth review of @tmssngr as well as you, @DieterMai, for addressing the comments!

I have tested the change on Windows and the toolbars work decently fine for me, so it's more than a good starting point for another replaced widget. Unfortunately, it does not work at all for me on MacOS (I will add a screenshot). Unfortunately, I have no immediate idea of what goes wrong on MacOS, so we will probably require some deeper analysis 🙁 

Regarding the known limitations of the toolbar implementation: I have just created an umbrella issue for collecting known limitations of widget, where we should then directly add the ones for toolbar to, so that someone will eventually address them:
- https://github.com/swt-initiative31/prototype-skija/issues/128

> Somehow if the background is transparent, skija does not provide a consistent result when calling surface.makeImageSnapshot(). Anyway, i fixed it by drawing a background with the background color.

That's interesting as the SkijaGC actually also clears the canvas upon creation with a background color. But by accident I just experienced the same issue when selecting a different size for buttons in the ControlExample. We have some general issues with the background anyway, so I think we can accept it as is.

> When i rebased on master, i realized the custom widgets now extend CustomControl. ToolBar can not do this. ToolBar can be a Parent for other widgets an thus needs to extend Composite for now.

That's fine. Once we have more composites custom-implemented, we can also start extracing things into common superclasses.

DieterMai commented:
@HeikoKlare how would you suggest to proceed here? I could of course somehow try setup a vm for mac. However i never done this so it might take a while.

HeikoKlare commented:
I just found out that CoolBar (an existing, emulated widget) is affected in the same way:
<img width="1800" alt="image" src="https://github.com/user-attachments/assets/04b84947-3dad-4821-a798-7d07fd1eb221" />

Since the CoolBar is working fine on upstream master, we must have made a change that broke the behavior and I hope that it affects CoolBar and ToolBar in the same way. I will try to figure out which change caused the issue as soon as possible because I think it's easiest (and hopefully rather quick to) find the cause when having a Mac at hand.

HeikoKlare commented:
Small update: unfortunately, I was wrong. The issue is introduced by this PR and with `master` state both CoolBar and ToolBar are working fine on MacOS. The reason why CoolBar is not working is that it is embedded into a ToolBar.

Until now I found that there are probably two combined issues:
1. The `isDrawing()` method in `Toolbar` always returns `false`, resulting in the ToolBar never being drawn. Changing it to `true` or just removing it (as it's already implemented in `Control`) makes the toolbar visible. Is there any reason for overwriting that method?
2. The bounds calculation is somehow buggy. The toolbar always has bounds (10, 10, 64, 64). If I am not mistaken that is some default size.
<img width="300" alt="image" src="https://github.com/user-attachments/assets/4a4e1659-ecf5-4fcf-9c17-34f838b6c655" />

Both will probably be easy to fix. I will have a look at the second issues as well.

Btw: I like the interaction feedback we now have on MacOS. The native MacOS toolbar does not provide any feedback for hovering or clicking.

DieterMai commented:
> 1. The `isDrawing()` method in `Toolbar` always returns `false`, resulting in the ToolBar never being drawn. Changing it to `true` or just removing it (as it's already implemented in `Control`) makes the toolbar visible. Is there any reason for overwriting that method?

I don't know exactly anmore. I think i had to add this to have a unit test fixed for mac. But if it is in the Mac Control class i don't see why

> 2. The bounds calculation is somehow buggy. The toolbar always has bounds (10, 10, 64, 64). If I am not mistaken that is some default size.

Maybe we have to override computeSize(). The current implementation only overrides computeSizeInPixels() but this method does not even exist for mac. So seeing this, it seems obvious that it could not work for mac since the size is never computed.


> Btw: I like the interaction feedback we now have on MacOS. The native MacOS toolbar does not provide any feedback for hovering or clicking.

That's the advantage of having only one implementation. All features get merged.

Thank you for your support @HeikoKlare

HeikoKlare commented:
> Maybe we have to override computeSize(). The current implementation only overrides computeSizeInPixels() but this method does not even exist for mac. So seeing this, it seems obvious that it could not work for mac since the size is never computed.

That's exactly the issue I also found meanwhile :-)

The currently implemented `computeSizeInPixels` is not present in the MacOS widgets implementations, thus never called. Replacing it with `computeSize()` (and removing the `isDrawing()` method) does the trick:
<img width="619" alt="image" src="https://github.com/user-attachments/assets/ce6ed6b2-c6a2-4311-8553-bd38e15635e0" />
<img width="1102" alt="image" src="https://github.com/user-attachments/assets/a6af7101-4775-4be7-b59f-6192ecbd4f95" />


The missing/misplaced drop-down icons seem to be caused by not doing proper auto-scaling (as I just replaced `computeSizeInPixels` with `computeSize`).

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/ToolBar.java:
```diff
@@ -0,0 +1,592 @@
+	public Point computeSize(int wHint, int hHint, boolean changed) {
+		checkWidget();
+		int zoom = getZoom();
+		wHint = (wHint != SWT.DEFAULT ? DPIUtil.scaleUp(wHint, zoom) : wHint);
+		hHint = (hHint != SWT.DEFAULT ? DPIUtil.scaleUp(hHint, zoom) : hHint);
```
We should probably just use `DPIUtil.autoScaleUp()` (without a `zoom`) here for now. That's what we currently do everywhere, as the widget-specific zoom is something Windows-specific we currently do not support. Then the `getZoom()` methods in the Cocoa and GTK Widget implementation can be removed again (they would need to be implemented differently anyway).

Or probably using any kind of scaling is even not necessary/correct here due to the calculations being performed in points anyway (as explained in the comment below), so we may completely remove the scaleUp call here.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/ToolBar.java:
```diff
@@ -0,0 +1,592 @@
+		int zoom = getZoom();
+		wHint = (wHint != SWT.DEFAULT ? DPIUtil.scaleUp(wHint, zoom) : wHint);
+		hHint = (hHint != SWT.DEFAULT ? DPIUtil.scaleUp(hHint, zoom) : hHint);
+		return DPIUtil.scaleDown(computeSizeInPixels(wHint, hHint, changed), zoom);
+	}
+
+	Point computeSizeInPixels(int wHint, int hHint, boolean changed) {
```
If I am not mistaken, these lines have to be removed. The calculation done by the renderer is performed in points, thus treating the result as pixels leads to wrong results. Simply treating it as points (without any scaling in between) seem to work fine for me.
```suggestion
		return DPIUtil.scaleDown(computeSizeInPixels(wHint, hHint, changed), zoom);
	}

	Point computeSizeInPixels(int wHint, int hHint, boolean changed) {
```

HeikoKlare reviewed (approved):
Thank you for the great contribution, @DieterMai!

Could you please document known limitations (e.g., stretching separators on MacOS) in the umbrella issue (https://github.com/swt-initiative31/prototype-skija/issues/128) so that we do not forget about them?

Currently, the toolbar will cause a segfault when running on Wayland. It will be fixed via https://github.com/swt-initiative31/prototype-skija/pull/131, which is why I propose to process that other PR as soon as possible and then merge this on top of it.
